docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+
+## Setup
+
+The `setup` method is often used like the constructor hook (`__init__`) in normal Python classes. However, for more advanced use cases it’s good to realize that it is not quite the same as a constructor.
+
+`setup` is only called after a Module becomes bound. Normally, this is not an issue because most Modules are bound (almost) immediately (as part of `init` and `apply`). Inside `setup`, sub-modules become bound when they are assigned to an attribute. nside an `nn.compact` decorated method, sub-modules are bound immediately when constructed. As explained in the previous section, top-level Modules are never bound and thus setup is not called when they are constructed. This means that you cannot access attributes assigned in setup from an unbound, top-level module.
+
+The `setup` method is not called immediately after the `Module` becomes bound but only when you interact with the `Module` instance (e.g.: call a method or access an attribute). This should not impact the behavior of a `Module` but the lazy execution does sometimes affect log statements and stack traces during debugging. The section on functionalization will explain why we need `setup` to be lazy in the first place.
+
+
+# Functionalization
+
+So far we had a pure `apply` function that is typically transformed with some JAX transformations and inside `apply` we have a stateful Module instance to work with. In other words: Outside of a Module we are in a functional world where we have the power of JAX's functional transformations and inside the Module we get the power of Flax's stateful variables and PRNG sequence, and the `apply` method is our bridge between these two worlds.
+
+But what if we want to use JAX transformations **inside** Modules? The answer to this is functionalization.
+
+This procedure itself is tedious and error-prone but handled internally by Flax. At a high-level we can summarize it as follows. For a method `fn` defined within a Module:
+
+1. Collect the state (variables & PRNG sequences) of the Module(s) that should be available inside the JAX transformation and take a snapshot of it.
+2. Call the JAX transformation with the original arguments and the collected state. Then inside the transformation:
+    1. Unpack the state and recreate the Modules
+    2. Call the user code `fn`
+    3. Collect the updated variables and rng and return it together with the original return values from `fn`
+3. Update the original state with the updated state returned from the transformation.
+
+A more in depth explanation of functionalization and lifting can be found [here](https://flax.readthedocs.io/en/latest/design_notes/lift.html).
```

Comment:
```suggestion
A more in depth explanation of functionalization and lifting can be found in the [Lifted Transformations](https://flax.readthedocs.io/en/latest/design_notes/lift.html) design note.
```

(This may be "better" than "here")

Reply:
+1

also consider adding a link back to this document from that other guide

Reply:
+1

also consider adding a link back to this document from that other guide